[Ruby] Change color of %i and %I symbol array literal elements according to symbol color (#1879)

themilkman opened the pull request:
This is my first try to implement #1879. I checked that Rubymine behaves in the same way.

An open task would be to adopt changes into `syntax_test_ruby.rb`, but I'd do this only after someone acknowledged the way I did it and that it's okay.

Examples would be:
```
%I(asd foo#{1 + 1 + 'x'} bar)
%I[asd foo#{1 + 1 + 'x'} bar]
%I<asd foo#{1 + 1 + 'x'} bar>
%I!asd foo#{1 + 1 + 'x'} bar!

%i(asd foo bar)
%i[asd foo bar]
%i<asd foo bar>
%i!asd foo bar!
```
Since I haven't worked on syntax files yet I would appreciate any feedback.

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
+          pop: true
+        - match: '\\\]|\\\\'
+          scope: constant.character.escape.ruby
+        - include: nest-bracketsa
```
This context name appears to contain a typo

```suggestion
        - include: nest-brackets
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
+          scope: constant.character.escape.ruby
+        - include: nest-ltgt
+    # symbol literal incapable of interpolation -- []
+    - match: '%[i]\['
```
```suggestion
    - match: '%i\['
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
+          scope: constant.character.escape.ruby
+        - include: nest-parens
+    # symbol literal incapable of interpolation -- <>
+    - match: '%[i]\<'
```
```suggestion
    - match: '%i\<'
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
         # Cant be named because its not necessarily an escape
         - match: \\.
+    # symbol literal incapable of interpolation -- ()
+    - match: '%[i]\('
```
```suggestion
    - match: '%i\('
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1030,8 +1030,72 @@ contexts:
         - include: interpolated-ruby
         - include: escaped-char
+    # literal capable of interpolation ()
+    - match: '%[I]?\('
```
An above rule matching `[QWS]?` will match before `I?`, so the `?` isn't useful. The square brackets aren't necessary since there aren't multiple letters that should be possible options.

```suggestion
    - match: '%I\('
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1030,8 +1030,72 @@ contexts:
+        - include: escaped-char
+        - include: nest-parens-i
+    # "literal capable of interpolation []"
+    - match: '%[I]?\['
```
```suggestion
    - match: '%I\['
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1030,8 +1030,72 @@ contexts:
+        - include: escaped-char
+        - include: nest-brackets-i
+    # literal capable of interpolation <>
+    - match: '%[I]?\<'
```
```suggestion
    - match: '%I\<'
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1030,8 +1030,72 @@ contexts:
+        - include: escaped-char
+        - include: nest-ltgt-i
+    # literal capable of interpolation -- {}
+    - match: '%[I]?\{'
```
```suggestion
    - match: '%I\{'
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1030,8 +1030,72 @@ contexts:
+        - include: escaped-char
+        - include: nest-curly-i
+    # literal capable of interpolation -- wildcard
+    - match: '%[I]([^\w])'
```
```suggestion
    - match: '%I([^\w])'
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
+          scope: constant.character.escape.ruby
+        - include: nest-bracketsa
+    # symbol literal incapable of interpolation -- {}
+    - match: '%[i]\{'
```
```suggestion
    - match: '%i\{'
```

wbond commented on Ruby/Ruby.sublime-syntax:
```diff
@@ -1084,6 +1148,70 @@ contexts:
+          scope: constant.character.escape.ruby
+        - include: nest-curly
+    # symbol literal incapable of interpolation -- wildcard
+    - match: '%[i]([^\w])'
```
```suggestion
    - match: '%i([^\w])'
```

themilkman commented:
Thank you very much @wbond, I'll take a deep look into this and happily resolve the issues. Also I'll try to understand how the tests work/get them locally working if possible (now that I see/guess what the mentionend `syntax_test_ruby.rb` file is good for ;-) ).

themilkman commented:
Hi @wbond , thanks again for your feedback. I adopted it and added some more tests. If you could review it and it is okay I'd squash the commits into one for a merge. Otherwise I'd happily do further work on it. In that case please give me some advice what to do. Anyway, thanks in advance!

themilkman commented:
Fyi, I just squashed the commits and rebased them against the current master. As always, I'd be glad to receive any further feedback or would be happy to see this change merged.

Thanks in advance.

themilkman commented:
Hi @wbond, I just rebased it again against the current master. Is there anything left which prevents the changes to be merged? I use "my" changes in everyday work and could not notice any problems with it yet (but really like the advantages).
I really know you're very busy but I'd welcome any feedback on this.

wbond commented:
Sorry, apparently I had a number of review comments but I didn't "finish" the review. I'm looking over now. If the changes are minor I'll merge this in and make the tweaks after.
